Move the client lib to use the standard lib/pq

andreimatei opened the pull request:
... instead of our forked one. This will help people who are trying to
link in both crdb and lib/pq.

The test fix was necessary because our forked libpq has a different way
of consuming results, making exhausting the `rows` cursor somehow not
necessary.

Fixes #17

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach-go/18)
<!-- Reviewable:end -->

tamird commented on crdb/tx_test.go:
```diff
@@ -22,7 +22,7 @@ import (
 	"sync"
 	"testing"
 
-	"github.com/cockroachdb/cockroach-go/testserver"
```
this ain't right

tamird commented on crdb/tx_test.go:
```diff
@@ -50,15 +50,18 @@ INSERT INTO d.t (acct, balance) VALUES (1, 100), (2, 100);
-				return
-			}
-			if err = rows.Scan(bal); err != nil {
+		defer rows.Close()
```
i don't understand what you changed here. can you help?

in particular, it seems that if you tried to scan a third value you'd hit an index out of bounds exception.

spencerkimball replied:
What Tamir said.

spencerkimball commented:
LGTM

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[crdb/tx.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-KRfOpQc2A852G4ngzw4:-KRfOpQc2A852G4ngzw5:binsxlw) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/c355fae646b95bd2a887080a2b6ea5b2a3a7c478/crdb/tx.go#L24)):*
> ```Go
> 	"database/sql"
> 
> 	"github.com/lib/pq"
> ```

Doesn't this prevent us from using `github.com/cockroachdb/pq` with `cockroach-go`?  I think some of the code in `github.com/cockroachdb/examples-go` will break. And some code in the main cockroach repo which uses `cockroach-go`.

---

*[crdb/tx_test.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-KRfPK72NfhCPFg5ds89:-KRfPK72NfhCPFg5ds8A:b2s2ria) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/c355fae646b95bd2a887080a2b6ea5b2a3a7c478/crdb/tx_test.go#L25)):*
> ```Go
> 	"testing"
> 
> 	"github.com/cockroach-go/testserver"
> ```

What is this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[crdb/tx.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-KRfOpQc2A852G4ngzw4:-KRgEc6_NHUJ4_9Ayk89:b-u8q5zw) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/c355fae646b95bd2a887080a2b6ea5b2a3a7c478/crdb/tx.go#L24)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Doesn't this prevent us from using `github.com/cockroachdb/pq` with `cockroach-go`?  I think some of the code in `github.com/cockroachdb/examples-go` will break. And some code in the main cockroach repo which uses `cockroach-go`.
</details>
yeah, the examples I was fixing.
I didn't realize that the main repo uses crdb too, but indeed it does for some acceptance test. And that sucker transitively depends on `github.com/cockroachdb/pq`. I think I could break that dependency, but I think the better thing to do is to remove all references to `github.com/cockroachdb/pq` from the main repo, except for the one in the CLI. Nobody should need our patches, except the CLI, and nobody depends on the CLI. What do you think?

Btw, I think I once used some tool to give an import chain between two packages, like a recursive `go list`, but now I can't find it. Do y'all happen to have a way?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[crdb/tx.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-KRfOpQc2A852G4ngzw4:-KRgGSfQoFkdnJuZ1Wcb:bmf2vvi) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/c355fae646b95bd2a887080a2b6ea5b2a3a7c478/crdb/tx.go#L24)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> yeah, the examples I was fixing.
> I didn't realize that the main repo uses crdb too, but indeed it does for some acceptance test. And that sucker transitively depends on `github.com/cockroachdb/pq`. I think I could break that dependency, but I think the better thing to do is to remove all references to `github.com/cockroachdb/pq` from the main repo, except for the one in the CLI. Nobody should need our patches, except the CLI, and nobody depends on the CLI. What do you think?
> 
> Btw, I think I once used some tool to give an import chain between two packages, like a recursive `go list`, but now I can't find it. Do y'all happen to have a way?
</details>
This was incomplete. We'd also need to make `github.com/cockroachdb/pq` register the driver under a different name than `postgres`, and have the CLI use that other name for querying.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[crdb/tx.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-KRfOpQc2A852G4ngzw4:-KRhmWQLa1u-WL_6fIH1:bi5qpnt) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/c355fae646b95bd2a887080a2b6ea5b2a3a7c478/crdb/tx.go#L24)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> This was incomplete. We'd also need to make `github.com/cockroachdb/pq` register the driver under a different name than `postgres`, and have the CLI use that other name for querying.
</details>
I agree with that plan. Only the cli needs to use `cockroachdb/pq`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[crdb/tx.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-KRfOpQc2A852G4ngzw4:-KRiemR76aHFMgcSXaLu:b-rec3i1) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/c355fae646b95bd2a887080a2b6ea5b2a3a7c478/crdb/tx.go#L24)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I agree with that plan. Only the cli needs to use `cockroachdb/pq`.
</details>
I went even further - there's no need for our fork to register any name (since the CLI doesn't use it through go's sql package). So I'm making that registration gated on a build tag, and should only be useful for running the library's tests.

So the plan is: land this and the fix to `go-examples` at the same time. Main cockroaches glocks this dependency, so it'll continue to run. Then land the change to main cockroach together with a glock update.

---

*[crdb/tx_test.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-KRfPK72NfhCPFg5ds89:-KRidcm3EaC2DmW6UHxJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/c355fae646b95bd2a887080a2b6ea5b2a3a7c478/crdb/tx_test.go#L25)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What is this?
</details>
Done.

---

*[crdb/tx_test.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-KRef0vJ0a6YpPNMqAgv-r1-53:-KRidr-NRnEkZH7vLjyE:b-2jwxsp) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/c355fae646b95bd2a887080a2b6ea5b2a3a7c478/crdb/tx_test.go#L53)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> What Tamir said.
</details>
i changed two things:

1) added the `defer rows.Close()`. This is mandated by the interface, and actually required with the unforked lib/pq, to consume all the results from the socket. Otherwise the connection gets desynchronized. This was sufficient to get the test passing.

2) the change below is just a minor rewrite. I think it's more idiomatic to iterate until `Next()` returns false. It's more defensive too - if the code had been written this way, it would've worked even without the `rows.Close()` call. And also it panics if the query suddenly starts returning 3 values - that's a good thing; think of it as a sanity assertion. Before, we would've silently ignored the 3rd value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[crdb/tx_test.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-KRef0vJ0a6YpPNMqAgv-r1-53:-KRihVaI_wwgFrqbW_Bi:b9ei572) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/c355fae646b95bd2a887080a2b6ea5b2a3a7c478/crdb/tx_test.go#L53)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> i changed two things:
> 
> 1) added the `defer rows.Close()`. This is mandated by the interface, and actually required with the unforked lib/pq, to consume all the results from the socket. Otherwise the connection gets desynchronized. This was sufficient to get the test passing.
> 
> 2) the change below is just a minor rewrite. I think it's more idiomatic to iterate until `Next()` returns false. It's more defensive too - if the code had been written this way, it would've worked even without the `rows.Close()` call. And also it panics if the query suddenly starts returning 3 values - that's a good thing; think of it as a sanity assertion. Before, we would've silently ignored the 3rd value.
</details>
I like the old version better. Just add an assertion that `rows.Next()` returns false after the `for .. := range` loop and call it a day.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/18#-:-KRj0LQZOBEtfWCpS57D:bnfp4nl)*
<!-- Sent from Reviewable.io -->
